### ISSUE TITLE
Deadtime correction in NRB calculation

### ISSUE THREAD

**Issue by bochens** [peterkuma/mpl2nc#1]
I am comparing the NRB calculated using this code to the NRB calculated using the SigmaMPL2015R2.3 version of the MPL software. I am getting a little higher result than the NRB given by the MPL software. I suspect that the dead time correction calculation might contribute to some differences. 

Is there a reason why in the dead time correction, the raw data is multiplied by 1000 when applied in the polynomial in the dead time correction?

It seems like with the dead time correction file that I was supplied for the lidar I am using, I had to time the input by 1000 to get a similar curve to Figure 5 in Campbell's 2002 paper. Is this the reason why there is a 1000 factor in this code?

**Comment by peterkuma** [comment]
My understanding is that the dead time correction coefficients as supplied in the vendor deadtime bin files are to be used with backscatter in units of Kcount/s. The units of the raw backscatter data are count/µs (= Mcount/s). That is why the raw data is multiplied by 1000 before applying the polynomial coefficients.

Here is an example dead time correction file from the vendor: [MMPL5054_SPCM34184_Deadtime7.zip](https://github.com/peterkuma/mpl2nc/files/12750057/MMPL5054_SPCM34184_Deadtime7.zip). The coefficients there are `[5.8747909e-30, -5.7959958e-25,  2.2592217e-20, -4.3401598e-16, 4.3104752e-12, -1.9502748e-08,  6.3204126e-05,  9.9846190e-01]`. When applied to input data `x` between 0 and 35 with the equation `np.sum([(x*1e3)**(n-i-1)*coeff[i] for i in range(n)], axis=0)` (as in mpl2nc), you get correction factors between about 1 and 12. This is consistent with the calibration curve supplied by the vendor ([System Information MiniMPL 5054.pdf](https://github.com/peterkuma/mpl2nc/files/12750109/System.Information.MiniMPL.5054.pdf)), where the x-axis values go from 0 to 35 Mcount/s.

It is possible that for your instrument the vendor dead time calibration file has a different format or units. There is also an issue with some vendor calibration files which contain wrong coefficients due to truncation of the float32 data type used in their file format.

Can you maybe provide some more information like an example backscatter file from your instrument, the vendor calibration files, and the "System Information" document for your instrument (if you have one)?

**Comment by bochens** [comment]
Looks like there might be some issue with my dead time correction file:
My deadtime coefficients are: [ 1.4012985e-45, -2.0838990e-40,  1.4304537e-35, -5.5098378e-31,
         1.3074610e-26, -1.9760829e-22,  1.9008108e-18, -1.1282178e-14,
         3.8515274e-11, -6.3340671e-08,  7.4575044e-05,  1.0080438e+00]

Here is the correction curve generated with these coefficients. The correction looks a little too high.
![ed001920-9ebf-424a-88dc-0c67adf3705f](https://github.com/peterkuma/mpl2nc/assets/105825473/137d7a3d-e95c-4145-822d-28f3d07cbf48)

Here is the calibration curve on the system information, and it does not match the curve in the dead time correction file.
<img width="864" alt="Screenshot 2023-09-28 at 9 46 27 AM" src="https://github.com/peterkuma/mpl2nc/assets/105825473/a8ac73ec-9e32-43bc-bb42-c317a9f7c743">

I also plotted your dead time correction file with my code, and the result looks like the system information provided by the vendor.

Here is my dead time correction file: 
[MMPL5051_SPCM34394_Deadtime11.bin.zip](https://github.com/peterkuma/mpl2nc/files/12750810/MMPL5051_SPCM34394_Deadtime11.bin.zip)


Added comments: although my dead time correction curve does not match the calibration curve in the system information at higher count rates, my calculated correction factor at 5 Mc/s and 10Mc/s does match the value shown in the system information document (1.24 at 5 Mc/s and 1.60 at 10 Mc/s). So, it is possible that my deadtime correction file isn't entirely wrong and the correction focus more on the smaller count rate. However, this will be an issue to discuss with the vendor, and the reason for the x1000 factor has been answered.

**Comment by peterkuma** [comment]
I think that the most likely reason is because the coefficients in your file are truncated to 32-bit floating point, which does not have enough precision to store the coefficients of longer polynomials. You could re-calculate the coefficients based on polynomial fitting (or a lookup table) of the curve in the system information.

We notified the vendor about the truncation issue about 3 years ago, but I am not sure if they have fixed it.

**Comment by bochens** [comment]
Thanks for the information. I will certainly create my own dead time correction.

Although we use an earlier miniMPL (unit 5051), I have only recently started trying to read the binary .mpl file and calculate the NRB. Your mpl2nc repo greatly inspires me. It is definitely better to take the data correction and analysis into our own hands instead of relying on the vendor software. I used and modified some of your code in my repo https://github.com/bochens/MPyLab that read MPL files as a data class, and I followed the same method for calculating NRB. I am currently working on adding functionalities like the Fernald retrieval method and cloud boundary detection to that repo.
